src/config/mod.rs

```diff
@@ -3571,44 +3564,42 @@ impl TikvConfig {
         config::canonicalize_sub_path(data_dir, DEFAULT_ROCKSDB_SUB_DIR)
     }
 
-    pub fn validate(&mut self) -> Result<(), Box<dyn Error>> {
-        // Setting up data paths.
-        if self.cfg_path.is_empty() {
-            self.cfg_path = Path::new(&self.storage.data_dir)
-                .join(LAST_CONFIG_FILE)
-                .to_str()
-                .unwrap()
-                .to_owned();
-        }
-
-        match (
-            self.raft_store.raftdb_path.is_empty(),
-            self.raft_engine.config.dir.is_empty(),
-        ) {
-            (false, true) => {
-                // If raftdb_path is specified, raft_engine_path will inherit it, this will be
-                // useful when updating from older version.
-                self.raft_engine.config.dir =
-                    self.infer_raft_engine_path(Some(self.raft_store.raftdb_path.as_str()))?;
-                self.raft_store.raftdb_path = self.infer_raft_db_path(None)?;
-            }
-            _ => {
-                self.raft_store.raftdb_path = self.infer_raft_db_path(None)?;
-                self.raft_engine.config.dir = self.infer_raft_engine_path(None)?;
+    fn titan_data_exists(&self) -> Result<bool, Box<dyn Error>> {
+        let exist = match self.storage.engine {
+            EngineType::RaftKv => {
+                let kv_db_path = self.infer_kv_engine_path(None)?;
+                // Check blob file dir is empty when titan is disabled
+                let titandb_path = if self.rocksdb.titan.dirname.is_empty() {
+                    Path::new(&kv_db_path).join("titandb")
+                } else {
+                    Path::new(&self.rocksdb.titan.dirname).to_path_buf()
+                };
+                if let Err(e) =
+                    tikv_util::config::check_data_dir_empty(titandb_path.to_str().unwrap(), "blob")
+                {
+                    if let Some(false) = self.rocksdb.titan.enabled {
+                        // If Titan is disabled explicitly but Titan's data directory is not empty,
+                        // return an error.
+                        return Err(format!(
+                            "check: titandb-data-dir-empty; err: \"{}\"; \
+                    hint: You have disabled titan when its data directory is not empty. \
+                    To properly shutdown titan, please enter fallback blob-run-mode and \
+                    wait till titandb files are all safely ingested.",
+                            e
+                        )
+                        .into());
+                    }
+                    false
```

Comment:
should it be false instead? as check_data_dir_empty returns error,  so it's not empty and `exist` should be true.